src/libraries/System.Text.Json/src/Resources/Strings.resx

```diff
@@ -506,10 +506,13 @@
   <data name="BufferMaximumSizeExceeded" xml:space="preserve">
     <value>Cannot allocate a buffer of size {0}.</value>
   </data>
-  <data name="CannotSerializeOpenGeneric" xml:space="preserve">
-    <value>The type '{0}' has generic parameters that have not been replaced by specific types, which is not valid during serialization or deserialization.</value>
+  <data name="CannotSerializeInvalidType" xml:space="preserve">
+    <value>The type '{0}' contains generic parameters that have not been replaced by specific types, is a pointer type, or is a ref struct. This is invalid during serialization or deserialization.</value>
```

Comment:
I do like the single message here vs. having 3 separate.

Reply:
I don't know about that. Having a single message is simpler for us as internal implementation detail, but it isn't super helpful to the end user. If I have a complex object graph and I see this exception message, how successful will I be in:
1) Reading the whole error message (or ignoring it because it is too long/seems complicated)
2) From that, figuring out which of the 2-3 cases I am in and what I need to do to fix my model.

If we are able to discern within the throwhelper/etc. what error state the caller is precisely in, we may want to give them an accurate/actionable error message rather than have the user try to figure out which one it is from a list of possible options.

That said, I wouldn't try to jump through too many hopes or pass a lot of state information around to achieve that. If it is relatively simple to do, we should do it, since I can see it add value to the customer.

Reply:
Specific error messages can help users. I'll revisit this during upcoming work. If we identify more classes of invalid types, we'll definitely have to split this up.

Reply:
Having separate messages for specific cases makes sense in many cases and we do it all over the place in the serializer, but is it worth the cost:benefit in this case? Cost is measured in initial implementation + testing overhead of course but also having to add strings to maintain (if the message is too specific), cost to localize (if we end up doing that), tiny hit on resource assembly\resource string size, and of course whether we can actually test the behavior.

If we can't reasonably test the exception string to make sure the correct error message is being used (without using literal text that might be localized), then I prefer a single string.